polkadot/node/core/pvf/common/src/worker/security.rs

```diff
@@ -17,30 +17,136 @@
+///
+/// NOTE: This should not be called in a multi-threaded context. `unshare(2)`:
+///
+///       > CLONE_NEWUSER requires that the calling process is not threaded.
+#[cfg(target_os = "linux")]
+pub fn unshare_user_namespace_and_change_root(
+	worker_kind: WorkerKind,
+	worker_dir_path: &std::path::Path,
+) -> Result<(), String> {
+	use std::{ffi::CString, os::unix::ffi::OsStrExt, ptr};
 
-/// To what degree landlock is enabled. It's a separate struct from `RulesetStatus` because that is
-/// only available on Linux, plus this has a nicer name.
-pub enum LandlockStatus {
-	FullyEnforced,
-	PartiallyEnforced,
-	NotEnforced,
-	/// Thread panicked, we don't know what the status is.
-	Unavailable,
+	let worker_dir_path_c = CString::new(worker_dir_path.as_os_str().as_bytes())
+		.expect("on unix; the path will never contain 0 bytes; qed");
+	let root_c = CString::new("/").expect("input contains no 0 bytes; qed");
+	let dot_c = CString::new(".").expect("input contains no 0 bytes; qed");
+
+	// Wrapper around all the work to prevent repetitive error handling.
+	//
+	// # Errors
+	//
+	// It's the caller's responsibility to call `Error::last_os_error`. Note that that alone does
+	// not give the context of which call failed, so we return a &str error.
+	|| -> Result<(), &'static str> {
+		// SAFETY: We pass null-terminated C strings and use the APIs as documented. In fact, steps
+		//         (2) and (3) are adapted from the example in pivot_root(2), with the additional
+		//         change described in the `pivot_root(".", ".")` section.
+		unsafe {
+			// 1. `unshare` the user and the mount namespaces.
+			if libc::unshare(libc::CLONE_NEWUSER | libc::CLONE_NEWNS) < 0 {
+				return Err("unshare user and mount namespaces")
+			}
+
+			// 2. Setup mounts.
+			//
+			// Ensure that new root and its parent mount don't have shared propagation (which would
+			// cause pivot_root() to return an error), and prevent propagation of mount events to
+			// the initial mount namespace.
+			if libc::mount(
+				ptr::null(),
+				root_c.as_ptr(),
+				ptr::null(),
+				libc::MS_REC | libc::MS_PRIVATE,
+				ptr::null(),
+			) < 0
+			{
+				return Err("mount MS_PRIVATE")
+			}
+			if libc::chdir(worker_dir_path_c.as_ptr()) < 0 {
+				return Err("chdir to worker dir path")
+			}
+			// Ensure that the new root is a mount point.
+			let additional_flags =
+				if let WorkerKind::Execute = worker_kind { libc::MS_RDONLY } else { 0 };
+			if libc::mount(
+				dot_c.as_ptr(),
+				dot_c.as_ptr(),
+				ptr::null(), // ignored when MS_BIND is used
+				libc::MS_BIND |
+					libc::MS_REC | libc::MS_NOEXEC |
+					libc::MS_NODEV | libc::MS_NOSUID |
+					libc::MS_NOATIME | additional_flags,
+				ptr::null(), // ignored when MS_BIND is used
+			) < 0
+			{
+				return Err("mount MS_BIND")
+			}
+
+			// 3. `pivot_root` to the artifact directory.
+			if libc::syscall(libc::SYS_pivot_root, &dot_c, &dot_c) < 0 {
```

Comment:
@koute help!!! I just noticed that this started failing after moving to `pivot_root(".", ".")`. The error is 22 - invalid argument. I tried also with `dot_c.as_ptr()`. Reverted to using an explicit `oldroot` dir for now.

Reply:
Ah, right, this shouldn't have worked; these must be pointers, not references to the pointers. :P

Remember that `libc::syscall` doesn't check in any way what you pass to it, so you *must* always manually make sure that its arguments are correct.

Also, there's no real need to put these in variables, and especially in a `CString` which will pointlessly allocate. You can just do `b".\0".as_ptr()` for `dot_c`, etc.

Reply:
Yeah, I tried with `as_ptr()` for both and it didn't help.

Reply:
Try changing your MS_BIND `mount` call to use absolute paths instead of `.`, then it should work I think.

Reply:
That worked, thanks!

Reply:
> I just tried `b".\0".as_ptr()`, but the pointer type is wrong (found `*const u8`, expected `*const i8`)

Yes, this is because the `libc` wrappers use `c_char` which can sometimes be signed and sometimes unsigned depending on the platform. It's fine to `.cast()` them.

Alternatively you could just do what I do, and copy-paste this somewhere in the utils:

```rust
// The following was copied from the `cstr_core` crate.
//
// TODO: Remove this once this is stable: https://github.com/rust-lang/rust/issues/105723
#[inline]
#[doc(hidden)]
#[allow(clippy::indexing_slicing)]
pub const fn cstr_is_valid(bytes: &[u8]) -> bool {
    if bytes.is_empty() || bytes[bytes.len() - 1] != 0 {
        return false;
    }

    let mut index = 0;
    while index < bytes.len() - 1 {
        if bytes[index] == 0 {
            return false;
        }
        index += 1;
    }
    true
}

#[macro_export]
macro_rules! cstr {
    ($e:expr) => {{
        const STR: &[u8] = concat!($e, "\0").as_bytes();
        const STR_VALID: bool = $crate::cstr_is_valid(STR);
        let _ = [(); 0 - (!(STR_VALID) as usize)];
        #[allow(unused_unsafe)]
        unsafe {
            core::ffi::CStr::from_bytes_with_nul_unchecked(STR)
        }
    }}
}
```

And then you can just do `cstr!(".").as_ptr()`.

But you know, whatever. :P At least use `CStr` instead of `CString` if you must.

Reply:
I'll use the macro, thanks!